### ISSUE TITLE
How to fetch different sections in same file in a single request

### ISSUE THREAD

**Issue by debugpoint136** [GMOD/tabix-js#34]
With a given file handle, if I want to fetch data line of A, B and C, I have to do 
`await tbiIndexed.getLines('A', 0, 1, (line, fileOffset) => lines.push(line)); `
( 0,1 because there is only 1 row for A )

 I need 2 more requests for B and C. Can this be done in a single request?

**Comment by cmdcolin** [comment]
You could make a wrapper interface that simply wraps to do multiple calls like this

```
getMultipleTabixRegions(arr /* can be ['A','B','C'] */, lineCallback) {
    return Promise.all(arr.map(name => tbiIndexed.getLines(name, 0, 1, lineCallback)))
}
```

If that is different from what you are asking let me know

I understand maybe you are using tabix-js as a name index also. That is pretty interesting and I'm curious if that is working as expected? Note: We have considered the possibility of making a trix index parser in JS which I think might enable some stuff like what you are doing now with tabix-as-name-store https://genome.ucsc.edu/goldenPath/help/trix.html

**Comment by debugpoint136** [comment]
I noted slightly better performance with your solution, file handle is created once instead of three times. But it still is 3 separate API requests. 
If I have a 2D matrix [1, 2, 3] by [A, B, C] it will make 9 requests. By using the wrapper the performance is slightly better, but doesn't scale too well. (my use case has matrix of 1000 x 1000)

I am indeed using tabix as name store, for example `(A)n	0	1	{"chr":"chr10","start":6780955,"end":6780982,"length":27,"class":"Simple_repeat","family":"Simple_repeat","readsCount":1,"RPKM":6.223,"RPM":0.168}`

 and very intrigued by trix (I didn't know about this). 
That's sounds very exciting. Are you still considering making trix index parser?

**Comment by cmdcolin** [comment]
Hi @debugpoint136 
Sorry for the delay
If you can identify anything in this library that could possibly be optimized we are definitely open to it

We are also considering making a "trix-js", I am unsure if the benefits will really be that good in querying large amounts of name data. It will still end up having to send many requests if the names are distributed all across the "spectrum" (not correlated)

I would maybe consider a dynamic backend e.g. database because I am unsure if a static solution can get very good performance if it just has to launch so many queries, but if you are seeing even passable performance with this library that is quite impressive :) does 1000x1000 names= 1,000,000 names?

**Comment by cmdcolin** [comment]
Note: Fetching multiple sections of a file in the same request is something that the byte-range Range header is possible of doing but I am not sure how that works. We have the http-range-fetcher library https://github.com/rbuels/http-range-fetcher that is a bit complicated but it  aggregates multiple requests that are being sent and then turns them into chunks and then returns it to the individual fetch calls transparently. We considered having multiple byte ranges to this library but it is not there yet

**Comment by debugpoint136** [comment]
Thanks for the pointer on http-range-fetcher. I will examine it closely and see if it fits my use case, or for future consideration. For now, I have tweaked the data files to collate 1000 rows to a single row, that way I can fetch all the data in a single request (although it fetches way too much data that I would need). Unfortunately, a dynamic backend, or an API endpoint is not an option, due to the nature of this use case. Thank you so much for your responses @cmdcolin. Closing this now..
